pingcap/tidb
# *: refine DML stmt's slow log

### What problem does this PR solve? 

DML(insert/update/insert/replace)'s slow log only contain detail about cop and 2pc.

but miss point-get(kvGet) and batchChecker(for duplicate / replaice) cost.

and hard to answer "why my plan is point-get update but executed so slow?"

### What is changed and how it works?

- Record Snapshot Cost in slow log
- Refactor `ExecDetail` into `CopDetail`, `SnapshotDetail`, `CommitDetail`
- Print Cop/Snaphost/Commit information in different lines in slow log 
- Record `WaitLockTime` instead of `ResolveLockTime`, previous one includes the resolveLockTime and backoff time before trying resolve.

when execute a `select into t select from t2 on duplicate update b = values(b)`;

before

```
# Process_time: 2 Wait_time: 60 Backoff_time: 0.001 Request_count: 2 Total_keys: 10000 Process_keys: 20001 Prewrite_time: 0.000000002 Commit_time: 0.000000003 Get_commit_ts_time: 0.000000001 Commit_backoff_time: 0.000000005 Resolve_lock_time: 0.000000002 Local_latch_wait_time: 0.000000004 Write_keys: 3 Write_size: 4 Prewrite_region: 5 Txn_retry: 6
```

after

```
# Cop_process_time: 2 Cop_wait_time: 60 Cop_backoff_time: 0.001 Cop_request_count: 2 Cop_total_keys: 10000 Cop_process_keys: 20001 Cop_wait_lock_time: 0.00012
# Get_time: 0.000000001 Get_key: 3 Get_wait_lock_time: 0.000000001 Get_rpc_count: 2
# Prewrite_time: 0.000000002 Commit_time: 0.000000003 Get_commit_ts_time: 0.000000001 Commit_backoff_time: 0.000000005 Commit_wait_lock_time: 0.000000002 Local_latch_wait_time: 0.000000004 Write_keys: 3 Write_size: 4 Prewrite_region: 5 Txn_retry: 6
```

then we can know real cost of read phase(cop + batchcheck) in this insert into select on duplicate, and see how long it used to wait lock in read or write phase

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
```
run and see log
```

Code changes

 - impl

Side effects

 - n/a

Related changes

 - Need to cherry-pick to the release branch

Release note

 - refine DML stmt's slow log


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/13212)
